File: engine-core/src/contracts.rs
//! Typed mirrors of `contracts/*.schema.json` (all v1.0.0).
//!
//! The schema files are the source of truth. Keep these types in lockstep;
//! a breaking change means a new schema version (new `$id`), never mutation.
//! `deny_unknown_fields` mirrors the schemas' `additionalProperties: false`.
//!
//! Constraints the type system cannot carry (runtime-enforced):
//! - `Market::Polymarket` is schema-valid but must be rejected by every v1
//!   runtime component.
//! - `risk.max_position_pct` must not exceed the `guardrails.toml` cap.
//! - A promotion with `to_stage: live` must have a fully populated
//!   `human_approval` (both message ids + timestamp) — re-verify before acting.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum Market {
    UsEquities,
    /// Reserved for v2 — rejected by all v1 runtime components.
    Polymarket,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum Family {
    MsShift,
    Swing,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum Lifecycle {
    Research,
    Backtest,
    Paper,
    Live,
    Retired,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum Severity {
    Info,
    Warning,
    High,
    Critical,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct SignalSpec {
    /// Always "python" in v1.
    pub language: String,
    /// `module_path.py:ClassName` implementing the Signal protocol.
    pub entrypoint: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct RiskSpec {
    pub max_position_pct: f64,
    pub stop_loss_pct: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct Scorecard {
    pub sharpe_wf: Option<f64>,
    pub sortino_wf: Option<f64>,
    pub max_drawdown_bt: Option<f64>,
    pub sharpe_paper: Option<f64>,
    pub max_drawdown_paper: Option<f64>,
    pub pnl_live: Option<f64>,
    pub rank: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct StrategyManifest {
    pub schema_version: String,
    pub id: String,
    pub wiki_page: String,
    pub market: Market,
    pub family: Family,
    pub universe: Vec<String>,
    pub hypothesis: String,
    pub signal_spec: SignalSpec,
    pub risk: RiskSpec,
    pub lifecycle: Lifecycle,
    pub scorecard: Scorecard,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct DatePeriod {
    /// ISO 8601 date (YYYY-MM-DD).
    pub start: String,
    pub end: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct SlippageAssumptions {
    pub model: String,
    pub bps: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct DataSnapshot {
    pub parquet_path: String,
    /// `sha256:<64 hex chars>` of the parquet file.
    pub content_hash: String,
    pub source_feed: String,
    pub period: DatePeriod,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct BacktestMetrics {
    pub sharpe: f64,
    pub sortino: f64,
    pub max_drawdown_pct: f64,
    pub turnover: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct BacktestResult {
    pub schema_version: String,
    pub strategy_id: String,
    pub period: DatePeriod,
    pub metrics: BacktestMetrics,
    pub slippage: SlippageAssumptions,
    pub equity_curve_path: String,
    pub data_snapshot: DataSnapshot,
    pub passed_thresholds: bool,
    pub notes: String,
    /// RFC 3339 timestamp.
    pub generated_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct PaperMetrics {
    pub sharpe: f64,
    pub sortino: f64,
    pub max_drawdown_pct: f64,
    pub pnl_usd: f64,
    pub num_trades: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct PaperResult {
    pub schema_version: String,
    pub strategy_id: String,
    pub period: DatePeriod,
    pub metrics: PaperMetrics,
    pub slippage: SlippageAssumptions,
    pub equity_curve_path: String,
    pub data_snapshot: DataSnapshot,
    pub passed_thresholds: bool,
    pub notes: String,
    pub generated_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct HumanApproval {
    pub required: bool,
    pub telegram_msg_id: Option<i64>,
    pub confirmation_msg_id: Option<i64>,
    pub approved_at: Option<String>,
}

impl HumanApproval {
    /// A live promotion may only act on a complete two-step approval.
    pub fn is_complete(&self) -> bool {
        self.required
            && self.telegram_msg_id.is_some()
            && self.confirmation_msg_id.is_some()
            && self.approved_at.is_some()
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct Promotion {
    pub schema_version: String,
    pub id: String,
    pub strategy_id: String,
    pub from_stage: Lifecycle,
    pub to_stage: Lifecycle,
    pub evidence: Vec<String>,
    pub rationale: String,
    pub issued_at: String,
    pub human_approval: HumanApproval,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct Event {
    pub schema_version: String,
    pub id: String,
    pub source_agent: String,
    pub severity: Severity,
    pub kind: String,
    pub payload: serde_json::Value,
    pub requires_reply: bool,
    pub ts: String,
}

File: sandbox/backtest/contracts.py
"""Typed mirrors of contracts/*.schema.json (all v1.0.0).

The schema files are the source of truth. Keep these models in lockstep;
a breaking change means a new schema version (new $id), never mutation.
extra="forbid" mirrors the schemas' additionalProperties: false.

Constraints the models cannot carry (runtime-enforced):
- market "crypto" is enabled in v1 (24/7 spot pairs; engine.py annualizes
  with 365 sessions/year instead of us_equities' 252 — see
  engine.SESSIONS_PER_YEAR).
- market "polymarket" is schema-valid but must be rejected by every v1
  runtime component.
- risk.max_position_pct must not exceed the guardrails.toml cap.
- A promotion with to_stage "live" must have a fully populated
  human_approval (both message ids + timestamp) — re-verify before acting.
"""

from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

SCHEMA_VERSION = "1.0.0"

Market = Literal["us_equities", "crypto", "polymarket"]
Family = Literal["ms_shift", "swing"]
Lifecycle = Literal["research", "backtest", "paper", "live", "retired"]
Severity = Literal["info", "warning", "high", "critical"]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class SignalSpec(StrictModel):
    language: Literal["python"]
    entrypoint: str


class RiskSpec(StrictModel):
    max_position_pct: float = Field(gt=0, le=100)
    stop_loss_pct: float = Field(gt=0)
    stop_loss_cooldown_sessions: int = Field(default=0, ge=0)


class Scorecard(StrictModel):
    sharpe_wf: Optional[float]
    sortino_wf: Optional[float]
    max_drawdown_bt: Optional[float]
    sharpe_paper: Optional[float]
    max_drawdown_paper: Optional[float]
    pnl_live: Optional[float]
    rank: Optional[int] = Field(ge=1)


class StrategyManifest(StrictModel):
    schema_version: Literal["1.0.0"]
    id: str = Field(pattern=r"^[a-z0-9]+(-[a-z0-9]+)*$")
    wiki_page: str
    market: Market
    family: Family
    universe: list[str] = Field(min_length=1)
    hypothesis: str = Field(min_length=1)
    signal_spec: SignalSpec
    risk: RiskSpec
    lifecycle: Lifecycle
    scorecard: Scorecard


class DatePeriod(StrictModel):
    start: str
    end: str


class SlippageAssumptions(StrictModel):
    model: str
    bps: float = Field(ge=0)


class DataSnapshot(StrictModel):
    parquet_path: str
    content_hash: str = Field(pattern=r"^sha256:[0-9a-f]{64}$")
    source_feed: str
    period: DatePeriod


class BacktestMetrics(StrictModel):
    sharpe: float
    sortino: float
    max_drawdown_pct: float = Field(ge=0)
    turnover: float = Field(ge=0)


class BacktestResult(StrictModel):
    schema_version: Literal["1.0.0"]
    strategy_id: str = Field(pattern=r"^[a-z0-9]+(-[a-z0-9]+)*$")
    period: DatePeriod
    metrics: BacktestMetrics
    slippage: SlippageAssumptions
    equity_curve_path: str
    data_snapshot: DataSnapshot
    passed_thresholds: bool
    notes: str
    generated_at: str


class PaperMetrics(StrictModel):
    sharpe: float
    sortino: float
    max_drawdown_pct: float = Field(ge=0)
    pnl_usd: float
    num_trades: int = Field(ge=0)


class PaperResult(StrictModel):
    schema_version: Literal["1.0.0"]
    strategy_id: str = Field(pattern=r"^[a-z0-9]+(-[a-z0-9]+)*$")
    period: DatePeriod
    metrics: PaperMetrics
    slippage: SlippageAssumptions
    equity_curve_path: str
    data_snapshot: DataSnapshot
    passed_thresholds: bool
    notes: str
    generated_at: str


class HumanApproval(StrictModel):
    required: bool
    telegram_msg_id: Optional[int]
    confirmation_msg_id: Optional[int]
    approved_at: Optional[str]

    def is_complete(self) -> bool:
        """A live promotion may only act on a complete two-step approval."""
        return (
            self.required
            and self.telegram_msg_id is not None
            and self.confirmation_msg_id is not None
            and self.approved_at is not None
        )


class Promotion(StrictModel):
    schema_version: Literal["1.0.0"]
    id: str = Field(pattern=r"^[a-z0-9]+(-[a-z0-9]+)*$")
    strategy_id: str = Field(pattern=r"^[a-z0-9]+(-[a-z0-9]+)*$")
    from_stage: Lifecycle
    to_stage: Lifecycle
    evidence: list[str]
    rationale: str = Field(min_length=1)
    issued_at: str
    human_approval: HumanApproval


class Event(StrictModel):
    schema_version: Literal["1.0.0"]
    id: str = Field(pattern=r"^[a-z0-9]+(-[a-z0-9]+)*$")
    source_agent: str = Field(min_length=1)
    severity: Severity
    kind: str = Field(min_length=1)
    payload: dict[str, Any]
    requires_reply: bool
    ts: str
